src/Workspaces/CSharp/Portable/Extensions/ContextQuery/CSharpSyntaxContext.cs

```diff
@@ -255,11 +255,10 @@ public static CSharpSyntaxContext CreateContext_Test(SemanticModel semanticModel
             return CreateContextWorker(workspace: null, semanticModel: semanticModel, position: position, cancellationToken: cancellationToken);
         }
 
-        private new static bool IsWithinAsyncMethod()
+        private new static bool IsWithinAsyncMethod(SyntaxToken targetToken)
         {
-            // TODO: Implement this if any C# completion code needs to know if it is in an async 
-            // method or not.
-            return false;
+            var methodDeclaration = targetToken.GetAncestor<MethodDeclarationSyntax>();
+            return methodDeclaration != null && methodDeclaration.Modifiers.Any(SyntaxKind.AsyncKeyword);
```

Comment:
what about async local functions? lambdas? anonymous methods?

Reply:
Hmm, good point...

Reply:
please add tests for that too

Reply:
So, this is already broken for `Async Function` in VB, and since I based the C# version on that, it inherits the problem. Async local functions and anonymous delegates<sup>*</sup> seem to reset the `async`-ness (unless of course they too are `async`), whereas lambdas inherit the `async`-ness of the nearest Method, local function, anonymous delegate or async lambda. Or something like that.

It really feels like this info should come from the compiler, instead of trying replicate the calculation. In any case, this is enough work for a separate issue, and I'm not making it any worse, so #wontfix :-)

<sub><sup>*</sup> I didn't even know `async delegate` was a thing.</sub>

Reply:
@lorcanmooney 
> whereas lambdas inherit the async-ness of the nearest Method

Yes, I checked that too. It seems really weird that the behavior is different between lambdas / anonymous methods. Might be a bug?

With lambdas, either making the lambda itself or the containing method does the trick. At least that's what syntax highlgihting is telling me.

> In any case, this is enough work for a separate issue, and I'm not making it any worse, so #wontfix :-)

I agree this is a very minor bug, so you don't have to fix it. But I think the case inside a method is pretty minor as well. 😃

Reply:
@lorcanmooney just so you know, I filed an issue for the behavior with lambdas: #25410 because it is most likely a bug. That would make the fix easier than you thought, but it's ok if you still don't want to do it. I might do it later myself.